### PR TITLE
added switch case to detect blank pointers in keeper

### PR DESCRIPTION
**Pull request by alizahidraja**
## Description

Added switch case to detect blank pointers in the keeper.

closes: #2268

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`) (not applicable)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/core/keeper/keeper.go [hunk 1]
```diff
@@ -50,12 +50,28 @@ func NewKeeper(
...
+
+	// switch case to detect blank pointers
+	switch reflect.TypeOf(stakingKeeper).Kind() {
+	case reflect.Ptr:
+		if reflect.ValueOf(&stakingKeeper).IsZero() {
+			panic(fmt.Errorf("cannot initialize IBC keeper: empty staking keeper"))
+		}
+	default:
+		if reflect.ValueOf(stakingKeeper).IsZero() {
+			panic(fmt.Errorf("cannot initialize IBC keeper: empty staking keeper"))
```
Comment (crodriguezvega) [inline modules/core/keeper/keeper.go:62]: Could we maybe have a slight different error message, so that in case the error occurs is immediately obvious what the problem was (i.e. either empty value or empty pointer)?
```suggestion
			panic(fmt.Errorf("cannot initialize IBC keeper: empty staking keeper value"))
```

File: modules/core/keeper/keeper.go [hunk 2]
```diff
@@ -50,12 +50,28 @@ func NewKeeper(
...
 
 	// panic if any of the keepers passed in is empty
-	if reflect.ValueOf(stakingKeeper).IsZero() {
-		panic(fmt.Errorf("cannot initialize IBC keeper: empty staking keeper"))
+
+	// switch case to detect blank pointers
+	switch reflect.TypeOf(stakingKeeper).Kind() {
+	case reflect.Ptr:
+		if reflect.ValueOf(&stakingKeeper).IsZero() {
+			panic(fmt.Errorf("cannot initialize IBC keeper: empty staking keeper"))
```
Comment (crodriguezvega) [inline modules/core/keeper/keeper.go:58]:
```suggestion
			panic(fmt.Errorf("cannot initialize IBC keeper: empty staking keeper pointer"))
```

**Review by crodriguezvega — COMMENTED** [review]
Thanks for fixing this, @alizahidraja!

Could you also please try to add some tests for the empty pointer situations [here](https://github.com/cosmos/ibc-go/blob/main/modules/core/keeper/keeper_test.go#L77)?

File: modules/core/keeper/keeper.go [hunk 3]
```diff
@@ -50,12 +50,28 @@ func NewKeeper(
 	}
 
 	// panic if any of the keepers passed in is empty
-	if reflect.ValueOf(stakingKeeper).IsZero() {
```
Comment (damiannolan) [inline modules/core/keeper/keeper.go:53]: I think [`ValueOf`](https://pkg.go.dev/reflect#ValueOf) returns the zero value for `nil`, i.e. blank pointers.
Perhaps we could add some test cases before adapting this code to use the switch case.

**Review by damiannolan — COMMENTED** [review]
Thanks for your contributions @alizahidraja, much appreciated! 🙏 

I have some doubts over this code needing to change, but perhaps some additional tests could get to the bottom of it!

File: modules/core/keeper/keeper.go (hunk 3 above)
Comment (alizahidraja) [reply modules/core/keeper/keeper.go:53]: Hey @damiannolan! 

In the code from the [issue](https://github.com/cosmos/ibc-go/issues/2268) raised, the behavior of a pointer value was not as predicted with `ValueOf`

Will try to add additional test cases to confirm the need for switch case.

Another approach can be to add another `or` statement in the same line, but I thought switch case would have more readability plus a different error message for value & pointers.

**Comment by alizahidraja** [comment]
@crodriguezvega @damiannolan I added 2 new test cases and ran them on both, the switch case code and previous code; the previous code fail the new test cases with pointers

Please do have a look and lmk your thoughts

**Review by crodriguezvega — APPROVED** [review]
Thanks, @alizahidraja!

File: modules/core/keeper/keeper.go (hunk 3 above)
Comment (colin-axner) [reply modules/core/keeper/keeper.go:53]: Could we modify the if statement to be:
```go
if stakingKeeper == nil || reflect.ValueOf(stakingKeeper).IsZero() {
```

I don't really want to over complicate this code. It is just a defensive check to prevent the chain from starting if the app.go is misconfigured. We simply want to know if the arguments provided are empty. I feel like there should be simple way of doing this? Otherwise we should add a private function to perform the switch so it just reads:

```go
assertNotEmpty(stakingKeeper)
assertNotEmpty(upgradeKeeper)
```

Comment (alizahidraja) [reply modules/core/keeper/keeper.go:53]: I understand the concern but running the following code

```fmt.Println(stakingKeeper == nil)```

when given a pointer, returns `false`

I tried to solve it in a simpler way but failed. I can make a separate function in the same file that would check if an interface object(struct or pointer) is empty if you think that approach is better here

Comment (colin-axner) [reply modules/core/keeper/keeper.go:53]: I see the nil check won't catch the &stakingtypes.StakingKeeper{} issue. 

I think my preference would be to move the switch to a private assertNotEmpty function which returns an error which we panic on wrapping the name of the keeper:
```go
if isEmpty(stakingKeeper) {
    panic(fmt.Errorf("cannot initialize IBC keeper: empty staking keeper"))
}
```

**Review by colin-axner — COMMENTED** [review]
Thanks for picking this up @alizahidraja. It's a bit of a tricky issue so I appreciate your patience. See my [latest comment](https://github.com/cosmos/ibc-go/pull/2403/files#r983395304). I think it makes sense to add this to the bottom of the keeper.go file

**Comment by alizahidraja** [comment]
No worries @colin-axner, I'll make a function to check for that at the bottom

File: modules/core/keeper/keeper.go (hunk 3 above)
Comment (alizahidraja) [reply modules/core/keeper/keeper.go:53]: As @crodriguezvega mentioned, it's good to know if the error was due to value or pointer, if we make a function as you suggest, we won't be able to know the difference

Is that okay or do you suggest we think of something else?

**Comment by alizahidraja** [comment]
Added an isEmpty function at the bottom of the keeper

The final concern would be that the error message cannot specify if the keeper object is a struct or a pointer, dk if that is important, lmk what you guys think

**Review by colin-axner — APPROVED** [review]
Thanks @alizahidraja!

**Comment by colin-axner** [comment]
> The final concern would be that the error message cannot specify if the keeper object is a struct or a pointer, dk if that is important, lmk what you guys think

No, I think as you have it now is great. We mostly just want to panic if something is misconfigured so the chain developer knows to look into the code
